## Data Preparation and Exploration

This notebook is focused on preparing, cleaning and exploring a dataset that as been gathered at two solar power plants in India over a 34 day period

The goal is to apply this dataset to machine learning models to predict solar power plant output.


Points to look out for in a clean data:
* Making sure the data is presented in a way that makes sense with all the variables correct.
* Ensure that the variables make sense and are correct.
* Eliminating duplicate rows and columns as well as elimiating columns that provide redundant information.
* Dealing with missing data appropriately.

In [142]:
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import math

import warnings
warnings.filterwarnings("ignore")


### Preliminary Analysis
The power generation datasets are gathered at the inverter level - each inverter has multiple lines of solar panels attached to it. The sensor data is gathered at a plant level - single array of sensors optimally placed at the plant.

In [143]:
### Reading in the data
generation_dataset = pd.read_csv('Plant_1_Generation_Data.csv')
weather_dataset = pd.read_csv('Plant_2_Weather_Sensor_Data.csv')

In [144]:
generation_dataset.head(10)

DATE_TIME  PLANT_ID       SOURCE_KEY  DC_POWER  AC_POWER  \
0  15-05-2020 00:00   4135001  1BY6WEcLGh8j5v7       0.0       0.0   
1  15-05-2020 00:00   4135001  1IF53ai7Xc0U56Y       0.0       0.0   
2  15-05-2020 00:00   4135001  3PZuoBAID5Wc2HD       0.0       0.0   
3  15-05-2020 00:00   4135001  7JYdWkrLSPkdwr4       0.0       0.0   
4  15-05-2020 00:00   4135001  McdE0feGgRqW7Ca       0.0       0.0   
5  15-05-2020 00:00   4135001  VHMLBKoKgIrUVDU       0.0       0.0   
6  15-05-2020 00:00   4135001  WRmjgnKYAwPKWDb       0.0       0.0   
7  15-05-2020 00:00   4135001  ZnxXDlPa8U1GXgE       0.0       0.0   
8  15-05-2020 00:00   4135001  ZoEaEvLYb1n2sOq       0.0       0.0   
9  15-05-2020 00:00   4135001  adLQvlD726eNBSB       0.0       0.0   

   DAILY_YIELD  TOTAL_YIELD  
0          0.0    6259559.0  
1          0.0    6183645.0  
2          0.0    6987759.0  
3          0.0    7602960.0  
4          0.0    7158964.0  
5          0.0    7206408.0  
6          0.0    7028673.0  
7          0.0    6522172.0  
8          0.0    7098099.0  
9          0.0    6271355.0

In [145]:
generation_dataset.shape

(68778, 7)

Generation dataset contains 68778 rows and 7 columns

In [146]:
weather_dataset.head(10)

DATE_TIME  PLANT_ID       SOURCE_KEY  AMBIENT_TEMPERATURE  \
0  2020-05-15 00:00:00   4136001  iq8k7ZNt4Mwm3w0            27.004764   
1  2020-05-15 00:15:00   4136001  iq8k7ZNt4Mwm3w0            26.880811   
2  2020-05-15 00:30:00   4136001  iq8k7ZNt4Mwm3w0            26.682055   
3  2020-05-15 00:45:00   4136001  iq8k7ZNt4Mwm3w0            26.500589   
4  2020-05-15 01:00:00   4136001  iq8k7ZNt4Mwm3w0            26.596148   
5  2020-05-15 01:15:00   4136001  iq8k7ZNt4Mwm3w0            26.512740   
6  2020-05-15 01:30:00   4136001  iq8k7ZNt4Mwm3w0            26.494339   
7  2020-05-15 01:45:00   4136001  iq8k7ZNt4Mwm3w0            26.420410   
8  2020-05-15 02:00:00   4136001  iq8k7ZNt4Mwm3w0            26.401946   
9  2020-05-15 02:15:00   4136001  iq8k7ZNt4Mwm3w0            26.226078   

   MODULE_TEMPERATURE  IRRADIATION  
0           25.060789          0.0  
1           24.421869          0.0  
2           24.427290          0.0  
3           24.420678          0.0  
4           25.088210          0.0  
5           25.317970          0.0  
6           25.217193          0.0  
7           25.065062          0.0  
8           24.691469          0.0  
9           24.559481          0.0

In [147]:
weather_dataset.shape

(3259, 6)

Weather dataset contains 3259 rows and 6 columns

Formatting DATE_TIME columns and merging the two dataframes.

In [148]:
generation_dataset['DATE_TIME'] = pd.to_datetime(generation_dataset['DATE_TIME'])
weather_dataset['DATE_TIME'] = pd.to_datetime(weather_dataset['DATE_TIME'])

In [149]:
df_solar_data = pd.merge(generation_dataset.drop(columns = ['PLANT_ID']), weather_dataset.drop(columns = ['PLANT_ID', 'SOURCE_KEY']), on='DATE_TIME')

In [150]:
# A summary of each column, the datatypes and the number of non-null values 

df_solar_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68708 entries, 0 to 68707
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   DATE_TIME            68708 non-null  datetime64[ns]
 1   SOURCE_KEY           68708 non-null  object        
 2   DC_POWER             68708 non-null  float64       
 3   AC_POWER             68708 non-null  float64       
 4   DAILY_YIELD          68708 non-null  float64       
 5   TOTAL_YIELD          68708 non-null  float64       
 6   AMBIENT_TEMPERATURE  68708 non-null  float64       
 7   MODULE_TEMPERATURE   68708 non-null  float64       
 8   IRRADIATION          68708 non-null  float64       
dtypes: datetime64[ns](1), float64(7), object(1)
memory usage: 4.7+ MB


Convert 'SOURCE_KEY' to numerical type

In [151]:
from sklearn.preprocessing import *

encoder = LabelEncoder()

df_solar_data['SOURCE_KEY_NUMBER'] = encoder.fit_transform(df_solar_data['SOURCE_KEY'])

df_solar_data.drop(columns=['SOURCE_KEY'], inplace=True)

In [152]:
df_solar_data.sample(10).style.background_gradient(cmap='hot')

### Looking for null and missing values.

In [153]:
df_solar_data.isnull().sum()


DATE_TIME              0
DC_POWER               0
AC_POWER               0
DAILY_YIELD            0
TOTAL_YIELD            0
AMBIENT_TEMPERATURE    0
MODULE_TEMPERATURE     0
IRRADIATION            0
SOURCE_KEY_NUMBER      0
dtype: int64

### Checking for Duplicate rows/ redundant information

In [154]:
#Checking for duplicate rows
df_solar_data.duplicated().sum()

0

In [155]:
#Checking for duplicate columns
df_solar_data.T.duplicated()

DATE_TIME              False
DC_POWER               False
AC_POWER               False
DAILY_YIELD            False
TOTAL_YIELD            False
AMBIENT_TEMPERATURE    False
MODULE_TEMPERATURE     False
IRRADIATION            False
SOURCE_KEY_NUMBER      False
dtype: bool